src/Stack/Upgrade.hs

```diff
@@ -29,16 +48,182 @@ import           Stack.Types.Config
+
+    binaryOpts = BinaryOpts
+        <$> fmap T.pack (strOption
+              ( long "binary-platform"
+             <> help "Platform type for archive to download"
+             <> (maybe mempty (value . T.unpack) mDefPlatform)
+             <> showDefault))
+
+    sourceOpts = SourceOpts
+        <$> ((\fromGit repo -> if fromGit then Just repo else Nothing)
+                <$> switch
+                    ( long "git"
+                    <> help "Clone from Git instead of downloading from Hackage (more dangerous)" )
+                <*> strOption
+                    ( long "git-repo"
+                    <> help "Clone from specified git repository"
+                    <> value "https://github.com/commercialhaskell/stack"
+                    <> showDefault ))
+
+data BinaryOpts = BinaryOpts
+    { _boPlatform :: !T.Text
+    }
+    deriving Show
+data SourceOpts = SourceOpts
+    { _soRepo :: !(Maybe String)
+    }
+    deriving Show
+
+data UpgradeOpts = UpgradeOpts
+    { _uoBinary :: !(Maybe BinaryOpts)
+    , _uoSource :: !(Maybe SourceOpts)
+    }
+    deriving Show
+
 upgrade :: (StackM env m, HasConfig env)
         => ConfigMonoid
-        -> Maybe String -- ^ git repository to use
         -> Maybe AbstractResolver
         -> Maybe String -- ^ git hash at time of building, if known
+        -> UpgradeOpts
         -> m ()
-upgrade gConfigMonoid gitRepo mresolver builtHash =
+upgrade gConfigMonoid mresolver builtHash uo@(UpgradeOpts mbo mso) =
+    case (mbo, mso) of
+        -- FIXME It would be far nicer to capture this case in the
+        -- options parser itself so we get better error messages, but
+        -- I can't think of a way to make it happen.
+        (Nothing, Nothing) -> error "You must allow either binary or source upgrade paths"
+        (Just bo, Nothing) -> binary bo
+        (Nothing, Just so) -> source so
+        (Just bo, Just so) -> binary bo `catchAny` \e -> do
+            $logWarn "Exception occured when trying to perform binary upgrade:"
+            $logWarn $ T.pack $ show e
+            $logWarn "Falling back to source upgrade"
+
+            source so
+  where
+    binary bo = binaryUpgrade bo
+    source so = sourceUpgrade gConfigMonoid mresolver builtHash so
+
+binaryUpgrade
+  :: (StackM env m, HasConfig env)
+  => BinaryOpts
+  -> m ()
+binaryUpgrade (BinaryOpts platform) = do
+    $logInfo $ "Querying for archive location for platform: " <> platform
+    archiveURL <- liftIO $ findArchive platform
+
+    $logInfo $ "Downloading from: " <> archiveURL
+
+    config <- askConfig
+    let destFile = toFilePath (configLocalBin config </> $(mkRelFile "stack"))
+#if WINDOWS
+                        FP.<.> "exe"
+#endif
+        tmpFile = destFile FP.<.> "tmp"
+
+    liftIO $ do
+      case () of
+        ()
+          | ".tar.gz" `T.isSuffixOf` archiveURL -> handleTarball archiveURL tmpFile
+          | ".zip" `T.isSuffixOf` archiveURL -> error "FIXME: Handle zip files"
```

Comment:
While we _could_ add zip file support, that's IMO a very bad idea: we likely don't want to add a zip library dependency (I think the ones available on Hackage are fairly unstable), and shelling out to an external command will be painful. Worse yet, this will introduce a lot of code that is conditional for Windows only.

What do you think about adding a tarball release option for Windows in addition to the zip files?